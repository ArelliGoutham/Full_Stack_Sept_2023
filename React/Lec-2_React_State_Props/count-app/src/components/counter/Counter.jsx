/**
 *   What is state in React?
 *     -> A state is a built-in react object that is used to contain data about the component.
 *     -> A component's state can change over the time, 
 *        whenever the state is going to change then the component has to re-render.
 * 
 *     -> How to respond events in react and update the UI?
 *        - What is an event? -> Remember in DOM API-> just an action like clicking the button, enter the value in input field.

 *     -> What is hook in react?
          - a hook is function that allows you to use the state in a functional component.
          - useState() -> is a built-in hook that manages the state of a functional component.
                       -> useState doesnâ€™t return just a variable, t returns an array, where the first element is the state variable 
                          and the second element is a function to update the value of the variable.
          - There are 2 kinds of hooks available in React:
            1. built-in hooks: useState(), useEffect(), useMemo(), useCallack() etc
            2. custom hooks -> useFetch() - user defined hooks can also be created via prefix "use"
 * 
 * 
 */

import { useState } from "react";
import './counter.css';

const Counter = ({ counter }) => {



    const [count, setCount] = useState(counter); // const count = 0;

    const handleIncrement = () =>{
        // setCount(count + 1);
        setCount((prevCount) => {
            const newCount = prevCount + 1;
            getData(newCount);
            return newCount;
        });
    }

    const handleDecrement = () =>{
        setCount(count - 1);
    }

    const getData = (count) => {
        console.log(count);
    }

    return (
        <>
          <button onClick={ handleIncrement }>+</button>
          <p>Counter Value: {count} </p>
          <button onClick={ handleDecrement }>-</button>
        </>
    )
}

export default Counter;