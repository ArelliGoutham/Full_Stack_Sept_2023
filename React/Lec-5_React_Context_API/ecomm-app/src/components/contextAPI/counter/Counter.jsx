/****
 *    What is context API?
 *       -> It's way to share the data among nested components without the need to pass props 
 *          manually through each level of component tree (nesting of components).
 *       -> Use cases: 
 *            1.themes, 
 *            2. user authentication
 *            3. settings etc
 *    How to implement the context API?
 *       Step-1: Create the context 
 *           import { createContext } from 'react';
 * 
 *           const CounterContext = createContext();
 * 
 *           export default CounterContext;
 *       Step-2: Need to create a provider this provider wraps the 
 *               whole appilcation for passing the data to the particlular components.
 * 
 *               you need created Context and useContext()
 * 
 *            Refer: CounterProvider.jsx file 
 *       Step-3: Consume the context
 *             custom to consume the context, but we need use useContext(ContextName);
 * 
 *     What are the benefits you get from context?
 *       1. Global data sharing, you can share data across the components.
 *       2. Simpler state management, It manages the state without complex setup.
 *       3. Flexibility to update any component.
 *       4. It provides you single source of truth.
 *       5. Avoiding the re-rendering which improves performance of react app.
 *       6. Clean code you can write via Context API. 
 * 
 */

import { useCounter } from "../../../context/CounterProvider";

const Counter = () => {
   
    const { handleIncrement } = useCounter();
    console.log('Counter');

    return (
        <>
          <button style={{fontSize: '24px'}} onClick={handleIncrement}>Increment</button>
        </>
    )
}

export default Counter;

