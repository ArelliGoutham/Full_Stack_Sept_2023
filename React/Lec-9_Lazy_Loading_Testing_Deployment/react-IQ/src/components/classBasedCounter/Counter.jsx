import { Component, PureComponent } from 'react';

class Counter extends PureComponent {
    constructor(props){
      super(props);
      this.state = {
        count: 0,
        //x: 10,
      }

      this.handleIncrement = this.handleIncrement.bind(this);
    }

    handleIncrement(){
        this.setState(
            (prevState) => ({
                count: prevState.count + 1
            })
        );
    }

    componentDidMount() {
        console.log('Component gets mounted');
    }

    componentDidUpdate(prevProps, prevState) {
       console.log('components gets updated, Count:', this.state.count);
    }

    componentWillUnmount(){
        console.log('Component gets unmounted');
    }

    render() {
        return(
            <>
                <p>Count: {this.state.count}</p>
                <button onClick={this.handleIncrement}>Increment</button>
            </>
        )
    }
}

export default Counter;