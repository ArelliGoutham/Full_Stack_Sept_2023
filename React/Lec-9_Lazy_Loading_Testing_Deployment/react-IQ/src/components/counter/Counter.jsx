import React, { useState } from 'react';
import './counter.css';

const Counter = () => {
    const [count, setCount] = useState(0);

    const handleIncrement = () => {
       setCount(count+1);
    }

    const handleDecrement = () => {
        setCount(count-1);
    }

    return (
        <div className="container counter-wrapper">

            <h1 className='fancy-text'>Counter example</h1>

            <h2>Counter Example with functional Component</h2>
            <button className="btn" onClick={handleIncrement}>+</button>
            <p>Count: {count}</p>
            <button className="btn" onClick={handleDecrement}>-</button>
        </div>
    )
}


export default React.memo(Counter);