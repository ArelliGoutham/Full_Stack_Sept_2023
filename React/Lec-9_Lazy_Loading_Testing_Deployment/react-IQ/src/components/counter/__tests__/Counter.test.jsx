import React from 'react';
import {render, fireEvent, screen } from '@testing-library/react';
import Counter from '../Counter';

describe('Counter Component', () => {

    test('snapshot testing', ()=>{
        const { asFragment } = render(<Counter />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the intial count as 0 and button text', ()=>{
        render(<Counter />);

        const incrementBtnText = screen.getByText('+');
        const decrementBtnText = screen.getByText('-');
        const countText = screen.getByText('Count: 0');

        expect(incrementBtnText).toBeTruthy();
        expect(decrementBtnText).toBeTruthy();
        expect(countText).toBeTruthy();

    });

    test('increment btn increases the count value', ()=>{
        // here write the logic to test it
        render(<Counter />);

        const increBtn = screen.getByText('+');
        fireEvent.click(increBtn);

        const countText = screen.getByText('Count: 1');
        expect(countText).toBeTruthy();
    });


    test('decrement btn decreases the count value', ()=>{
        // here write the logic to test it -> HW
    });
});