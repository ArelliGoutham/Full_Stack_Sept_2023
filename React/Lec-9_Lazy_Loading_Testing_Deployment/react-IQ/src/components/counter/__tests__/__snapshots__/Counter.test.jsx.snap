// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="container counter-wrapper"
  >
    <h1>
      Counter example
    </h1>
    <h2>
      Counter Example with functional Component
    </h2>
    <button
      class="btn"
    >
      +
    </button>
    <p>
      Count: 0
    </p>
    <button
      class="btn"
    >
      -
    </button>
  </div>
</DocumentFragment>
`;
